registry/cache/cache.go

```diff
@@ -153,43 +158,59 @@ func (c *cache) get(service string) ([]*registry.Service, error) {
 
 		// cache results
 		c.Lock()
-		c.set(service, util.Copy(services))
+		c.set(domain, service, util.Copy(services))
 		c.Unlock()
 
 		return services, nil
 	}
 
 	// watch service if not watched
-	_, ok := c.watched[service]
-
-	// unlock the read lock
+	c.RLock()
+	var ok bool
+	if _, d := c.watched[domain]; d {
+		if _, s := c.watched[domain][service]; s {
+			ok = true
+		}
+	}
 	c.RUnlock()
 
 	// check if its being watched
 	if !ok {
 		c.Lock()
 
+		// add domain if not registered
+		if _, ok := c.watched[domain]; !ok {
+			c.watched[domain] = make(map[string]bool)
+		}
+
 		// set to watched
-		c.watched[service] = true
+		c.watched[domain][service] = true
 
 		// only kick it off if not running
-		if !c.running {
-			go c.run()
+		if !c.running[domain] {
+			go c.run(domain)
 		}
 
 		c.Unlock()
 	}
 
 	// get and return services
-	return get(service, cp)
+	return get(domain, service, services)
 }
 
-func (c *cache) set(service string, services []*registry.Service) {
-	c.cache[service] = services
-	c.ttls[service] = time.Now().Add(c.opts.TTL)
+func (c *cache) set(domain string, service string, srvs []*registry.Service) {
```

Comment:
Is this likely to be called concurrently by different threads? If so we need some locking here.

Reply:
I've made a bunch of improvements to the registry cache so this should now be resolved.